### PR TITLE
Fixed MatrixNormal shape handling

### PR DESCRIPTION
**Pull request by kc611**
Linked Issue: #4845 

So it turns out that the shape handling was a bit off in `MatrixNormal` as it didn't support `mu` broadcasting, also it'll be better to just remove the `1x1` error since it won't make much sense if we have it supported already.

### CODE REVIEW

**Comment by kc611** [comment]
So the reason I've drafted this is :

The `logp` for `MatrixNormal` can only support a `value` of `ndim`s 2 as of now (see failing tests as they try to set an initial `value` with a 4 dimensional variable in `logp`), do we want it to support higher `ndim`s or should we restrict it similar to how `KroneckerNormal` does.

https://github.com/pymc-devs/pymc3/blob/7e464e59bcb0adb28df94f379b3e8d4af12bd4d1/pymc3/distributions/multivariate.py#L1892-L1893

 Though I don't know how this restriction works or wether it's even supposed to be there. (It was present in `v3`'s `KroneckerNormal`)

**Comment by michaelosthege** [comment]
If a restriction is associated with a clear error message asking for contributions, I'd go with that rather than to invest disproportionate much dev time in implementing a feature without knowing if there's even any demand for it.

But the changes you made look rather simple - could it be that the test is just failing because of a simple mixup?

**Comment by kc611** [comment]
I think in this case it'll be better to give it at-least 1 extra dimensional support. Cause similar cases are observed in `MvNormal` (and on further introspection `KroneckerNormal` actually does the same). They actually perform operations on 1-D values and provide support for a extra `size` dimension. In `MatrixNormal` we have 2-dimensional values so with only 2-dim support `MatrixNormal` won't work for any `size` we pass. (And that is why specifically the tests with `size` argument fail.) 

That said, is it possible to somehow 'iterate' over an axis of `TensorVariable` (`value` in this instance) to apply a given set of operations on last two dimensions ?

**Comment by ricardoV94** [comment]
> That said, is it possible to somehow 'iterate' over an axis of `TensorVariable` (`value` in this instance) to apply a given set of operations on last two dimensions ? 

I don't know what exactly you have in mind but axis can be tuples such as `np.sum(arr, axis=(-2, -1))` 

Or `numpy.apply_over_axes`, but that might not exist in aesara (yet)

**Comment by kc611** [comment]
The exact thing I wanted to do is basically apply the operations in the current`MatrixNormal` logp method (which contain some `Op`s explicitly designed for 2-dimensional(matrix) kind of values ) over an third dimensional axis (which will represent the `size`) parameter. So it'll be like iteration in which we apply the current operations in all values over the `size` axis.

The `MvNormal` (whose values are one dimension lower than that of `MatrixNormal`) does a similar thing in its logp by having that `onedim` variable which allows it to process one dimension higher. (The size dimension) The problem here is, unlike `MatrixNormal` we can't just add in an extra dimension (or can we?) , since the `Op`s here explicitly check for 2 dimensional values and I'm pretty sure they do it for good reasons. 

Basically I want the `TensorVariable` equivalent of:

```
for _value in value:
      do_the_thing(_value)
```

File: pymc3/distributions/multivariate.py [hunk 1]
```diff
@@ -1646,16 +1650,21 @@ def dist(
...
         **kwargs,
     ):
 
         cholesky = Cholesky(lower=True, on_error="raise")
 
-        if mu.ndim == 1:
-            raise ValueError(
-                "1x1 Matrix was provided. Please use Normal distribution for such cases."
+        if "size" in kwargs:
+            raise NotImplementedError("MatrixNormal doesn't support size argument")
```
Comment (kc611) [inline pymc3/distributions/multivariate.py:1660]: Turns out `size` was never supported in the first place in `MatrixNormal` in `v3`. (Thanks to @ricardoV94 for pointing that out.)

Comment (ricardoV94) [reply pymc3/distributions/multivariate.py:1660]: Users may also pass shape (and invalid shapes) or "dims", so this check might not be enough. Can a ndim check be done inside the logp method instead? And what about the RV, does it work properly for non-default sizes?

Comment (ricardoV94) [reply pymc3/distributions/multivariate.py:1660]: Also just to be clear 2d shapes (rows x cols) were allowed, but not "replication" shapes. I am not sure if that maps to any non-`None` size because I don't really grasp what size means for multivariate distributions.

Comment (kc611) [reply pymc3/distributions/multivariate.py:1660]: I added a warning for `shape` since if only `(m x n)` shape is allowed we can directly derive `m` and `n` from parameters. 

> And what about the RV, does it work properly for non-default sizes?

What do you mean by non-default sizes?

Comment (ricardoV94) [reply pymc3/distributions/multivariate.py:1660]: If you used the RV directly with some size like (2, 3, 3), would it do a sensible thing in the `rng_fn`? Or should it also raise there?

Comment (kc611) [reply pymc3/distributions/multivariate.py:1660]: The RV can handle sizes like that because the main operation over there i.e. `np.matmul` will treat the random output (from `rng.standard_normal` ) as stack of matrices (which is what we intend) so I don't think we need to raise there.

Comment (ricardoV94) [reply pymc3/distributions/multivariate.py:1660]: I think size is one of the default arguments for distributions. Perhaps it's better to simply check it is `None`

**Comment by ricardoV94** [comment]
So the issue in #4845 was just the wrong specification of `size`?

We need tests that cover those new warnings / errors.

File: pymc3/distributions/multivariate.py [hunk 2]
```diff
@@ -1646,16 +1650,21 @@ def dist(
...
         *args,
         **kwargs,
     ):
 
         cholesky = Cholesky(lower=True, on_error="raise")
 
-        if mu.ndim == 1:
-            raise ValueError(
-                "1x1 Matrix was provided. Please use Normal distribution for such cases."
+        if "size" in kwargs and kwargs["size"] is not None:
```
Comment (ricardoV94) [inline pymc3/distributions/multivariate.py]:
```suggestion
        if kwargs.get("size", None) is not None:
```

File: pymc3/distributions/multivariate.py [hunk 3]
```diff
@@ -1707,6 +1721,9 @@ def logp(value, mu, rowchol, colchol):
         TensorVariable
         """
 
+        if value.ndim != 2:
+            raise ValueError("Value must be two dimensional.")
```
Comment (ricardoV94) [inline pymc3/distributions/multivariate.py:1726]: Need a test that hits this branch

File: pymc3/distributions/multivariate.py [hunk 4]
```diff
@@ -1646,16 +1650,21 @@ def dist(
...
 
         cholesky = Cholesky(lower=True, on_error="raise")
 
-        if mu.ndim == 1:
-            raise ValueError(
-                "1x1 Matrix was provided. Please use Normal distribution for such cases."
+        if "size" in kwargs and kwargs["size"] is not None:
+            raise NotImplementedError("MatrixNormal doesn't support size argument")
+
+        if "shape" in kwargs:
```
Comment (ricardoV94) [inline pymc3/distributions/multivariate.py:1662]: Should add a test for this warning. And it should be a DeprecationWarning

File: pymc3/distributions/multivariate.py [hunk 5]
```diff
@@ -1646,16 +1650,22 @@ def dist(
...
-        if mu.ndim == 1:
-            raise ValueError(
-                "1x1 Matrix was provided. Please use Normal distribution for such cases."
+        if kwargs.get("size", None) is not None:
+            raise NotImplementedError("MatrixNormal doesn't support size argument")
+
+        if "shape" in kwargs:
+            kwargs.pop("shape")
+            warnings.warn(
+                "Passing shape to MatrixNormal is now depricated."
```
Comment (ricardoV94) [inline pymc3/distributions/multivariate.py]:
```suggestion
                "The shape argument in MatrixNormal is deprecated and will be ignored."
```

**Review by ricardoV94 — APPROVED** [review]

**Comment by michaelosthege** [comment]
Wow, approved 20 days ago. Can we merge this?

**Comment by kc611** [comment]
Yeah, the work here is done as far as I know.
